bin/check-doc

```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
+
+# Copyright Materialize, Inc. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+#
+# check-doc - check the docs for valid links
+
+set -euo pipefail
+
+cd "$(dirname "$0")/.."
+source misc/shlib/shlib.bash
+
+export RUSTDOCFLAGS='-D warnings'
+
+# to find the next thing to fix add `-p materialized` to the below list, `cargo doc` will stop
+# compiling on the first dependency that has invalid docs.
+runv cargo doc \
+    -p mz-avro \
+    -p ccsr \
+    -p expr \
+    -p interchange \
+    -p ore \
+    -p transform \
+    -p sql \
+    ;
```

Comment:
Not a big fan of leaving this half finished! Let's just fix it all.

Reply:
I was planning on making continuous progress, the goal of this was to prevent regressions as that happened.

Reply:
Yep, understood; I've been trying to plumb `#![deny(clippy::as_conversions)]` around for like a year now. But I haven't really made much progress since the initial push. IME these low-urgency projects tend to get forgotten about quickly, so when there's the possibility of just getting it all done at once, it's worth making the extra push!